monitor/inclusion_checker.go

```diff
@@ -0,0 +1,195 @@
+		return fmt.Errorf("error getting current log index for %q: %s", ic.logURI, err)
+	}
+
+	certs, err := ic.db.GetUnseen(ic.logID)
+	if err != nil {
+		inclusionErrors.WithLabelValues("getUnseen").Inc()
+		return fmt.Errorf("error getting unseen certificates from %q: %s", ic.logURI, err)
+	}
+	if len(certs) == 0 {
+		// nothing to do, don't advance the index
+		return nil
+	}
+
+	sth, err := ic.client.GetSTH(context.Background())
+	if err != nil {
+		inclusionErrors.WithLabelValues("getSTH").Inc()
+		return fmt.Errorf("error getting STH from %q: %s", ic.logURI, err)
+	}
+	newHead, entries, err := ic.getEntries(current, int64(sth.TreeSize))
+	if err != nil {
+		inclusionErrors.WithLabelValues("getEntries").Inc()
+		return fmt.Errorf("error retrieving entries from %q: %s", ic.logURI, err)
+	}
+
+	err = ic.checkEntries(certs, entries)
+	if err != nil {
+		inclusionErrors.WithLabelValues("checkEntries").Inc()
+		return fmt.Errorf("error checking retrieved entries for %q: %s", ic.logURI, err)
+	}
+
+	err = ic.db.UpdateIndex(ic.logID, newHead)
+	if err != nil {
+		inclusionErrors.WithLabelValues("updateIndex").Inc()
+		return fmt.Errorf("error updating current index for %q: %s", ic.logURI, err)
+	}
+
+	return nil
+}
+
+func min(a, b int64) int64 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func (ic *inclusionChecker) getEntries(start, end int64) (int64, []ct.LogEntry, error) {
+	if ic.maxGetEntries > 0 && end-start > ic.maxGetEntries {
+		end = start + ic.maxGetEntries
+	}
+	var allEntries []ct.LogEntry
+	for start <= end {
+		batchEnd := min(start+ic.batchSize, end)
+		entries, err := ic.client.GetEntries(context.Background(), start, batchEnd)
+		if err != nil {
+			return 0, nil, err
+		}
+		allEntries = append(allEntries, entries...)
+		start += int64(len(entries))
+	}
+	return start, allEntries, nil
+}
+
+func (ic *inclusionChecker) checkEntries(certs []storage.SubmittedCert, entries []ct.LogEntry) error {
+	// Key structure for our lookup map is as follows: SHA256 hash of the certificate
+	// body concatenated with the byte encoding of the SCT timestamp. This prevents
+	// from having duplicate keys for duplicate submissions with differing SCTs.
+	lookup := make(map[[32]byte]storage.SubmittedCert)
+	for _, cert := range certs {
+		content := make([]byte, len(cert.Cert)+binary.MaxVarintLen64)
+		copy(content, cert.Cert)
+		binary.PutUvarint(content[len(cert.Cert):], cert.Timestamp)
+		lookup[sha256.Sum256(content)] = cert
+	}
+	for _, entry := range entries {
+		var content []byte
+		switch entry.Leaf.TimestampedEntry.EntryType {
+		case ct.X509LogEntryType:
+			content = entry.X509Cert.Raw
+		case ct.PrecertLogEntryType:
+			content = entry.Precert.Submitted.Data
+		}
+		timestampBuf := make([]byte, binary.MaxVarintLen64)
+		binary.PutUvarint(timestampBuf, entry.Leaf.TimestampedEntry.Timestamp)
+		content = append(content, timestampBuf...)
+		h := sha256.Sum256(content)
+		if matching, found := lookup[h]; found {
```

Comment:
Should there be an error stat incremented for the SCT unmarshaling, signature verification, and storage marking in this loop?

Reply:
This gets caught by the stats in `checkInclusion`, at which point I think it would be more useful to simply check the logs than have another error stat which won't actually provide much insight other than which part failed (but not why), either way the logs will need to be checked to understand what is going wrong.

Reply:
Makes sense :+1: